### PR TITLE
Add TraceFlags class

### PR DESCRIPTION
**Pull request by fbogsany**
This implement `TraceFlags`. This used to be called `TraceOptions`, but https://github.com/open-telemetry/opentelemetry-specification/pull/234 proposes a rename to match the W3C spec.

This also implements `generate_span_id` and `generate_trace_id` in `SpanContext`.

### CODE REVIEW

File: api/lib/opentelemetry/trace/trace_flags.rb [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    # TraceFlags contain details about the trace. Unlike Tracestate values,
+    # TraceFlags are present in all traces. Currently, the only TraceFlag is a
+    # boolean {sampled?} {https://www.w3.org/TR/trace-context/#trace-flags flag}.
+    class TraceFlags
+      class << self
+        private :new # rubocop:disable Style/AccessModifierDeclarations
+
+        # Returns a newly created {TraceFlags} with the specified flags.
+        #
+        # @param [Integer] flags 8-bit byte of bit flags
```
[the remark this answers is not in the document]
Comment (fbogsany) [reply api/lib/opentelemetry/trace/trace_flags.rb]: Nope - literally an 8-bit byte. I'm following the pattern from http://ruby-doc.org/core-2.6.3/ARGF.html#method-i-getbyte.

**Comment by fbogsany** [comment]
> Does it make sense to represent trace and span ids as lowercase hex internally?

It is certainly a reasonable choice. We cheated a little in Shopify's internal tracing project, and used lowercase hex trace IDs, but integers in the range `1..(2**62 - 1)` for span IDs. The latter are generated more often and can be stored more efficiently (small integer).

We can use `Random::DEFAULT.bytes(16).unpack1('H*')` to generate trace IDs, however we have to wrap this in a loop, checking for all 0.

File: api/lib/opentelemetry/trace/span_context.rb [hunk 2]
```diff
@@ -6,25 +6,40 @@
...
       end
+
+      private
+
+      SPAN_ID_RANGE = (1..(2**64 - 1)).freeze
+      TRACE_ID_RANGE = (1..(2**128 - 1)).freeze
+
+      private_constant(:SPAN_ID_RANGE, :TRACE_ID_RANGE)
+
+      def generate_trace_id
```
[the remark this answers is not in the document]
Comment (fbogsany) [reply api/lib/opentelemetry/trace/span_context.rb]: I considered this, but couldn't find a convenient and obvious place for them. We don't currently have any methods on the `OpenTelemetry::Trace` module, for example, and the `OpenTelemetry::Trace::Tracer` class is meant to be a no-op implementation. Should we make it part of the `Tracer` API, and allow alternative implementations to provide their own?

**Comment by mwear** [comment]
I think going with a string for our internal representation is a good approach for now. If it turns we regularly need both string and integer representations, we can consider #23.

**Comment by fbogsany** [comment]
Bumping this comment, in case it was missed.

> I considered this, but couldn't find a convenient and obvious place for them. We don't currently have any methods on the `OpenTelemetry::Trace` module, for example, and the `OpenTelemetry::Trace::Tracer` class is meant to be a no-op implementation. Should we make it part of the `Tracer` API, and allow alternative implementations to provide their own?

File: api/lib/opentelemetry/trace/span_context.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (mwear) [reply api/lib/opentelemetry/trace/span_context.rb]: My initial thought was to put them on `OpenTelemetry::Trace`. Are there any reservations other than there aren't any methods defined on it yet?

[the remark this answers is not in the document]
Comment (fbogsany) [reply api/lib/opentelemetry/trace/span_context.rb]: Not really. Overall, though, this turned out to be more annoying than expected, due to `require` ordering.

**Comment by fbogsany** [comment]
Added tests for `SpanContext` to exercise these codepaths, and fixed the bugs they uncovered.

File: api/lib/opentelemetry/trace/span_context.rb [hunk 3]
```diff
@@ -6,24 +6,32 @@
...
+        @trace_flags = trace_flags
       end
 
       def valid?
-        !(@trace_id.zero? || @span_id.zero?)
+        @trace_id != OpenTelemetry::Trace::INVALID_TRACE_ID &&
+          @span_id != OpenTelemetry::Trace::INVALID_SPAN_ID
+      end
+
+      def self.invalid
```
Comment (fbogsany) [inline api/lib/opentelemetry/trace/span_context.rb]: I'm not really happy about this. I'd prefer an `INVALID` or `DEFAULT` constant, but the `require` order makes this difficult (`new` tries to access `OpenTelemetry::Trace::TraceFlags::DEFAULT` which hasn't been defined yet).

File: api/lib/opentelemetry/trace/trace_flags.rb [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+        # @raise [ArgumentError] If flags is not an 8-bit byte
+        # @return [TraceFlags]
+        def from_byte(flags)
+          raise ArgumentError, 'flags must be an 8-bit byte' unless (0..255).include?(flags)
```
Comment (fbogsany) [inline api/lib/opentelemetry/trace/trace_flags.rb]: Is it more efficient to `unless flags & ~0xff == 0`?

**Comment by fbogsany** [comment]
This is ready for review.

File: api/lib/opentelemetry/trace/trace_flags.rb (hunk 4 above)
Comment (mwear) [reply api/lib/opentelemetry/trace/trace_flags.rb]: The bitwise check is a clear winner:

```ruby
require 'benchmark/ipsa'

num = 144

Benchmark.ipsa do |x|
  x.report 'range check' do
    (0..255).include?(num)
  end
  x.report 'bitwise check' do
    num & ~0xFF == 0
  end
  x.compare!
end
```

Results:

```
Allocations -------------------------------------
         range check       0/0  alloc/ret        0/0  strings/ret
       bitwise check       0/0  alloc/ret        0/0  strings/ret
Warming up --------------------------------------
         range check   180.625k i/100ms
       bitwise check   205.490k i/100ms
Calculating -------------------------------------
         range check      7.869M (± 4.7%) i/s -     39.376M
       bitwise check     15.487M (± 2.7%) i/s -     77.470M

Comparison:
       bitwise check: 15486958.1 i/s
         range check:  7868744.9 i/s - 1.97x slower
```

File: api/lib/opentelemetry/trace.rb [hunk 5]
```diff
@@ -4,9 +4,47 @@
...
 # SPDX-License-Identifier: Apache-2.0
 
+module OpenTelemetry
+  # The Trace API allows recording a set of events, triggered as a result of a
+  # single logical operation, consolidated across various components of an
+  # application.
+  module Trace
+    # An invalid trace identifier, a 16-byte array with all zero bytes, encoded
+    # as a hexadecimal string.
+    INVALID_TRACE_ID = '0' * 32
```
Comment (mwear) [inline api/lib/opentelemetry/trace.rb]: Should we freeze this string (and INVALID_SPAN_ID below)?

**Comment by ibawt** [comment]
can this be approved now?  It looks as though the comments have been addressed?

**Pull request merged**